File: javascript/ch1-array-and-strings/longestSubstringEqualNumber.js
// Given some string having number of  '*' and '#' as its only characters find the longest
// substring which has highest number of equal number of '*' and '#'.

const longestSubstringEqualNumber = (word) => {
  let length = 0;
  let cumDiff = 0;
  let firstIndex = { 0: -1 };
  let longestString = "";
  for (let i = 0; i < word.length; i++) {
    let char = word[i];
    if (char === "*") {
      cumDiff += 1;
    } else if (char === "#") {
      cumDiff -= 1;
    }
    if (cumDiff in firstIndex) {
      length = Math.max(length, i - firstIndex[cumDiff]);
      longestString = word.slice(i - length + 1, i + 1);
    } else {
      firstIndex[cumDiff] = i;
    }
    console.log({ firstIndex, longestString, cumDiff, length });
  }
  return longestString;
};

console.log(longestSubstring("********#*#****###")); // #*#****###
// console.log(longestSubstring("#*")); // #*
// console.log(longestSubstring("####****"));
// console.log(longestSubstring("*#*#"));

File: java/algos/src/main/java/com/leetcode/LongestSubstringEqualNumber.java
package com.leetcode;

import java.util.HashMap;
import java.util.Map;

// Given some string having number of '*' and '#' as its only characters find the longest
// substring which has the highest number of equal number of '*' and '#'.
public class LongestSubstringEqualNumber {
    public static String solution(String word) { // O(n) time | O(n) space
        int length = 0;
        int cumDiff = 0;
        Map<Integer, Integer> firstIdx = new HashMap<>(); // O(n) space
        firstIdx.put(0, -1);
        String longestString = "";
        for (int i = 0; i < word.length(); i++) { // O(n) time
            char c = word.charAt(i);
            if (c == '#') {
                cumDiff += 1;
            } else if (c == '*') {
                cumDiff -= 1;
            }
            if (firstIdx.containsKey(cumDiff)) {
                length = Math.max(length, i - firstIdx.get(cumDiff));
                longestString = word.substring(i - length + 1, i + 1);
            } else {
                firstIdx.put(cumDiff, i);
            }
        }
        return longestString;
    }
}
